### ISSUE TITLE
Provide path_derivative argument to all continuous distributions 

### ISSUE THREAD

**Issue by botev** [thu-ml/zhusuan#67]
Paper: https://arxiv.org/abs/1703.09194 
You can just add to any continuous distribution an extra argument `path_derivative`. When that is true when calculating the `_log_prob` just wrap the distribution parameters in `tf.stop_gradients`.

**Comment by thjashin** [comment]
Thanks @botev . This is good since it just keep existing code behaviors unchanged. I've thought about this sticking-landing estimator but was focusing on how to do it in the variational objective which requires change of the api. Do you have interest in submitting a PR? We should appreciate it.

**Comment by botev** [comment]
Sure, I'll be happy to make a PR. 
Not use what kind of tests would be needed though?

**Comment by thjashin** [comment]
A simple test that checks the gradient actually doesn't go through the log_p term in that way would be enough (e.g., log_p has no gradient w.r.t. parameters if givens are not samples). And if I'm right you only need to implement this for distributions that has a `is_reparameterized` argument (`Normal`, `Concrete`, etc.).

**Comment by botev** [comment]
Ok and lastly do you want to have `use_path_derivative` as an argument to `log_prob` and `_log_prob`, to be as an attribute in the general `Distribution` class (like `is_reparametrized`) or attribute in the concrete classes. I think it should be an attribute to `log_prob` and `_log_prob` in the abstract `Distribution` class.

**Comment by thjashin** [comment]
I feel it's good to have it as an attribute to the `Distribution` class but you may need to actually do the `stop_gradient` in each concrete class because their parameters vary, just like how the `is_reparameterized` argument is dealt with.
